### PR TITLE
net: lib: nrf_cloud_coap: expose CoAP transport functions

### PR DESCRIPTION
**Pull request by jayteemo**
Expose transport functions so that external coap_client instances can be used with nRF Cloud.
IRIS-8806

### CODE REVIEW

**Comment by jayteemo** [comment]
i didn't include a changelog entry for this PR since it is not updating the public API.
follow-up PRs will use these changes.

**Review by plskeggs — APPROVED** [review]

**Review by plskeggs — APPROVED** [review]

**Review by peknis — APPROVED** [review]

File: subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c [hunk 1]
```diff
@@ -50,12 +50,23 @@ LOG_MODULE_REGISTER(nrf_cloud_coap_transport, CONFIG_NRF_CLOUD_COAP_LOG_LEVEL);
...
+	coap_client_response_cb_t cb;
+	void *user_data;
+	int result_code;
+	struct k_sem *sem;
+};
+
+/* Semaphore to be used with internal coap_client requests */
+static K_SEM_DEFINE(cb_sem, 0, 1);
+/* Semaphore to be used when doing authorization with an external coap_client */
+static K_SEM_DEFINE(ext_cc_sem, 0, 1);
```
Comment (glarsennordic) [inline subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c:67]: Would it make more sense for `ext_cc_sem` to be a part of the nrf_cloud_coap_client struct?

File: subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c [hunk 2]
```diff
@@ -633,29 +526,326 @@ int nrf_cloud_coap_fetch(const char *resource, const char *query,
...
+
+	memcpy(&opt_proxy_uri->value[uri_idx], PROXY_URI_DL_SEP, PROXY_URI_DL_SEP_LEN);
+	uri_idx += PROXY_URI_DL_SEP_LEN;
+
+	memcpy(&opt_proxy_uri->value[uri_idx], path, path_len);
+
+	LOG_DBG("Proxy URI: %.*s", opt_proxy_uri->len, opt_proxy_uri->value);
+
+	return 0;
+}
```
Comment (glarsennordic) [inline subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c:850]: This seems like a whole brand new feature to me? I also don't see it used anywhere

Should this be in a different PR?

**Review by glarsennordic — APPROVED** [review]

File: subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c (hunk 2 above)
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c:850]: this is an API that will be used for proxy downloads.
a future PR will use it.

File: subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c (hunk 1 above)
Comment (jayteemo) [reply subsys/net/lib/nrf_cloud/coap/src/nrf_cloud_coap_transport.c:67]: this is just used for auth and, unless things change, there will only be one external coap client (for downloads).
